Guide Register field in admin panel (#610)
strapi/documentation: docs/developer-docs/latest/guides/registering-a-field-in-admin.md

## Plugin setup

1. Create a new project:

    :::: tabs card

    ::: tab yarn

    Create an application and prevent the server from starting automatically as we will create a plugin right after the project generation.

    ```
    yarn create strapi-app my-app --quickstart --no-run
    ```

    :::

    ::: tab npx

    Create an application and prevent the server from starting automatically as we will create a plugin right after the project generation.

    ```
    npx create-strapi-app@latest my-app --quickstart --no-run
    ```

    :::

    ::::

2. Generate a plugin:

    :::: tabs card

    ::: tab yarn

    ```
    cd my-app
    yarn strapi generate
    ```

    Choose "plugin" from the list, press Enter and name the plugin wysiwyg.

    :::

    ::: tab npm

    ```
    cd my-app
    npm run strapi generate
    ```
    Choose "plugin" from the list, press Enter and name the plugin wysiwyg.

    :::

    ::::

3. Enable the plugin by adding it to the [plugins configurations](/developer-docs/latest/setup-deployment-guides/configurations/optional/plugins.md) file:

    ```js
    // path: /my-app/config/plugins.js

    module.exports = {
      // ...

>       'my-plugin': {

Reviewer: Shouldn't the plugin name be `wysiwyg` ?

        enabled: true,
        resolve: './src/plugins/wysiwyg' // path to plugin folder
      },
      // ...
    }
    ```

4. Install the needed dependencies:

    :::: tabs card

    ::: tab yarn

    ```
    cd src/plugins/wysiwyg
    yarn add @ckeditor/ckeditor5-react @ckeditor/ckeditor5-build-classic
    ```

    :::

    ::: tab npm

    ```
    cd src/plugins/wysiwyg
    npm install @ckeditor/ckeditor5-react @ckeditor/ckeditor5-build-classic
    ```

    :::

    ::::

5. Build then start your application with the front-end development mode:

    :::: tabs card

    ::: tab yarn

    ```

>     # Go back to to strapi root folder

Reviewer: Go back to the application root folder.

    cd ../../..

>     yarn build

Reviewer: There's no need to build the admin at this step.

    yarn develop --watch-admin
    ```

    :::

    ::: tab npm

    ```
    # Go back to to strapi root folder
    cd ../../..

>     npm run build

Reviewer: Same here you can remove this step.

      </code-block>
    </code-group>

::: note NOTE

> To launch Strapi in watch mode without creating a user account first will open `Port: 1337` with a JSON format error. Creating a user on `Port: 8081` will remove this alert.

Reviewer 2: ```suggestion
Launching the Strapi server in watch mode without creating a user account first will open `localhost:1337` with a JSON format error. Creating a user on `localhost:8081` prevents this alert.
```

export default MediaLib;
```
:::

> ::: details Example of an Editor component using CKEditor which will be the implementation of the new WYSIWYG:

Reviewer 2: ```suggestion
::: details Example of an Editor component using CKEditor as the WYSIWYG editor:
```

  );
};

Editor.defaultProps = {
  value: '',
  disabled: false
};

Editor.propTypes = {
  onChange: PropTypes.func.isRequired,
  name: PropTypes.string.isRequired,
  value: PropTypes.string,
  disabled: PropTypes.bool
};

export default Editor;
```

:::

> ::: details Example of a Wysiwyg component which will wrap the CKEditor:

Reviewer 2: ```suggestion
::: details Example of a Wysiwyg component wrapping CKEditor:
```

          disabled={disabled} 
          name={name} 
          onChange={onChange} 
          value={value} 
        />
        {error && 
          <Typography variant="pi" textColor="danger600">
            {formatMessage({ id: error, defaultMessage: error })}
          </Typography>
        }
        {description && 
          <Typography variant="pi">
            {formatMessage(description)}
          </Typography>
        }
      </Stack>
      <MediaLib 
        isOpen={mediaLibVisible} 
        onChange={handleChangeAssets}
        onToggle={handleToggleMediaLib} 
      />
    </>
  );
};

Wysiwyg.defaultProps = {
  description: '',
  disabled: false,
  error: undefined,
  intlLabel: '',
  required: false,
  value: '',
};

Wysiwyg.propTypes = {
  description: PropTypes.shape({
    id: PropTypes.string,
    defaultMessage: PropTypes.string,
  }),
  disabled: PropTypes.bool, 
  error: PropTypes.string, 
  intlLabel: PropTypes.shape({
    id: PropTypes.string,
    defaultMessage: PropTypes.string,
  }),
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  required: PropTypes.bool,
  value: PropTypes.string, 
};

export default Wysiwyg;
```

:::

> ## Register the field

Reviewer 2: See my comment at line 15 about headings wording/consistency :-)

The last step is to register the field `wysiwyg` with the WYSIWYG component with `addFields()`.

```js
// path: wysiwyg/admin/src/index.js

import pluginPkg from "../../package.json";
import Wysiwyg from "./components/Wysiwyg";
import pluginId from "./pluginId";

const name = pluginPkg.strapi.name;

export default {
  register(app) {
    app.addFields({ type: 'wysiwyg', Component: Wysiwyg });

    app.registerPlugin({
      id: pluginId,
      isReady: true,
      name,
    });
  },
  bootstrap() {},
};
```

> And VOILA, if you create a new `collectionType` or a `singleType` with a `richtext` field you will see the implementation of [CKEditor](https://ckeditor.com/ckeditor-5/) instead of the default WYSIWYG.

Reviewer 2: ```suggestion
And _voilà_, if you create a new `collectionType` or a `singleType` with a `richtext` field you will see the implementation of [CKEditor](https://ckeditor.com/ckeditor-5/) instead of the default WYSIWYG.
```

It's OK to be a bit more friendly here, but we should avoid using caps :-) 
I've added italics since it's a French word in an English sentence.

import PropTypes from 'prop-types';
import { Stack } from '@strapi/design-system/Stack';
import { Box } from '@strapi/design-system/Box';
import { Button } from '@strapi/design-system/Button';
import { Typography } from '@strapi/design-system/Typography';
import Landscape from '@strapi/icons/Landscape';
import MediaLib from '../MediaLib';
import Editor from '../Editor';
import { useIntl } from 'react-intl';

const Wysiwyg = ({ name, onChange, value, intlLabel, disabled, error, description, required }) => {
  const { formatMessage } = useIntl();
  const [mediaLibVisible, setMediaLibVisible] = useState(false);

  const handleToggleMediaLib = () => setMediaLibVisible(prev => !prev);

  const handleChangeAssets = assets => {
    let newValue = value ? value : '';

    assets.map(asset => {
      if (asset.mime.includes('image')) {
        const imgTag = `<p><img src="${asset.url}" alt="${asset.alt}"></img></p>`;

        newValue = `${newValue}${imgTag}`
      }

      // Handle videos and other type of files by adding some code
    });

    onChange({ target: { name, value: newValue } });
    handleToggleMediaLib();
  };
  
  return (
    <>
      <Stack size={1}>
        <Box>

>           <Typography variant="pi" fontWeight="bold">

Reviewer: `<Typography as="label" htmlFor={name} /... />`

>             {formatMessage({ id: intlLabel.id, defaultMessage: intlLabel.defaultMessage })}

Reviewer: `{formatMessage(intlLabel)}`

          </Typography>
        }
        {description && 
          <Typography variant="pi">
            {formatMessage({ id: description.id, defaultMessage: description.defaultMessage })}
          </Typography>
        }
      </Stack>
      <MediaLib 
        isOpen={mediaLibVisible} 
        onChange={handleChangeAssets}
        onToggle={handleToggleMediaLib} 
      />
    </>
  );
};

Wysiwyg.defaultProps = {
  description: '',
  disabled: false,
  error: undefined,
  intlLabel: '',
  required: false,
  value: '',
};

Wysiwyg.propTypes = {
  description: PropTypes.shape({
    id: PropTypes.string,
    defaultMessage: PropTypes.string,
  }),
  disabled: PropTypes.bool, 
  error: PropTypes.string, 
  intlLabel: PropTypes.shape({
    id: PropTypes.string,
    defaultMessage: PropTypes.string,
  }),
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  required: PropTypes.bool,
  value: PropTypes.string, 
};

export default Wysiwyg;
```

:::

## Register the field

The last step is to register the field `wysiwyg` with the WYSIWYG component with `addFields()`.

```js
// path: wysiwyg/admin/src/index.js

> import Wysiwyg from './components/Wysiwyg';

Reviewer: ```js
import pluginPkg from "../../package.json";
import Wysiwyg from "./components/Wysiwyg";
import pluginId from "./pluginId";

const name = pluginPkg.strapi.name;

export default {
  register(app) {
    app.addFields({ type: "wysiwyg", Component: Wysiwyg });

    app.registerPlugin({
      id: pluginId,
      isReady: true,
      name,
    });
  },
};
```
